common/src/main/java/io/netty/util/ResourceLeakDetectorFactory.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util;
+
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.lang.reflect.Constructor;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+/**
+ * This static factory should be used to load ResourceLeakDetectors as needed
+ */
+public abstract class ResourceLeakDetectorFactory {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(ResourceLeakDetectorFactory.class);
+
+    private static volatile ResourceLeakDetectorFactory INSTANCE = new ResourceLeakDetectorFactory() {
+        @Override
+        public <T> ResourceLeakDetector<T> loadResourceLeakDetector(final Class<T> resource) {
+            return AccessController.doPrivileged(new PrivilegedAction<ResourceLeakDetector<T>>() {
+                @Override
+                public ResourceLeakDetector<T> run() {
+                    final String customLeakDetector = SystemPropertyUtil.get("io.netty.customResourceLeakDetector");
+                    try {
```

Comment:
The loading code can't be in a static block because it needs a resource class to initialize the detector and (ideally) a generic type provided by the instance method. However, loading the system property can happen statically at the top of the class.